src/mono/mono/metadata/icall-eventpipe.c

```diff
@@ -491,6 +522,101 @@ eventpipe_execute_rundown (
 	return TRUE;
 }
 
+static
+gboolean
+eventpipe_walk_managed_stack_for_thread_func (
+	MonoStackFrameInfo *frame,
+	MonoContext *ctx,
+	gpointer data)
+{
+	g_assert_checked (frame != NULL);
+	g_assert_checked (data != NULL);
+
+	switch (frame->type) {
+	case FRAME_TYPE_DEBUGGER_INVOKE:
+	case FRAME_TYPE_MANAGED_TO_NATIVE:
+	case FRAME_TYPE_TRAMPOLINE:
+	case FRAME_TYPE_INTERP_TO_MANAGED:
+	case FRAME_TYPE_INTERP_TO_MANAGED_WITH_CTX:
+		return FALSE;
+	case FRAME_TYPE_MANAGED:
+	case FRAME_TYPE_INTERP:
+		if (!frame->ji)
+			return FALSE;
+		MonoMethod *method = frame->ji->async ? NULL : frame->actual_method;
+		ep_stack_contents_append ((EventPipeStackContents *)data, (uintptr_t)((uint8_t*)frame->ji->code_start + frame->native_offset), method);
+		return ep_stack_contents_get_length ((EventPipeStackContents *)data) >= EP_MAX_STACK_DEPTH;
+	default:
+		g_assert_not_reached ();
+		return FALSE;
+	}
+}
+
+static
+gboolean
+eventpipe_walk_managed_stack_for_thread (
+	ep_rt_thread_handle_t thread,
+	EventPipeStackContents *stack_contents)
+{
+	g_assert_checked (thread != NULL);
+	g_assert_checked (stack_contents != NULL);
+
+	if (thread == ep_rt_thread_get_handle ())
+		mono_get_eh_callbacks ()->mono_walk_stack_with_ctx (eventpipe_walk_managed_stack_for_thread_func, NULL, MONO_UNWIND_SIGNAL_SAFE, stack_contents);
+	else
+		mono_get_eh_callbacks ()->mono_walk_stack_with_state (eventpipe_walk_managed_stack_for_thread_func, mono_thread_info_get_suspend_state (thread), MONO_UNWIND_SIGNAL_SAFE, stack_contents);
+
+	return TRUE;
+}
+
+static
+gboolean
+eventpipe_method_get_simple_assembly_name (
+	ep_rt_method_desc_t *method,
+	ep_char8_t *name,
+	size_t name_len)
+{
+	g_assert_checked (method != NULL);
+	g_assert_checked (name != NULL);
+
+	MonoClass *method_class = mono_method_get_class (method);
+	MonoImage *method_image = method_class ? mono_class_get_image (method_class) : NULL;
+	const ep_char8_t *assembly_name = method_image ? mono_image_get_name (method_image) : NULL;
+
+	if (!assembly_name)
+		return FALSE;
+
+	size_t assembly_name_len = strlen (assembly_name) + 1;
+	size_t to_copy = assembly_name_len < name_len ? assembly_name_len : name_len;
+	memcpy (name, assembly_name, to_copy);
+	name [to_copy - 1] = 0;
+
+	return TRUE;
+}
+
+static
+gboolean
+evetpipe_method_get_full_name (
+	ep_rt_method_desc_t *method,
+	ep_char8_t *name,
+	size_t name_len)
+{
+	g_assert_checked (method != NULL);
+	g_assert_checked (name != NULL);
+
+	char *full_method_name = mono_method_get_name_full (method, TRUE, TRUE, MONO_TYPE_NAME_FORMAT_IL);
+	if (!full_method_name)
+		return FALSE;
+
+	size_t full_method_name_len = strlen (full_method_name) + 1;
+	size_t to_copy = full_method_name_len < name_len ? full_method_name_len : name_len;
+	memcpy (name, full_method_name, to_copy);
+	name [to_copy - 1] = 0;
```

Comment:
could use `g_strlcpy` potentially

Reply:
Did it this way to not bother about number of characters to copy (since it is utf8), so just retrieved length (will actually be number of bytes to copy) and then copy the whole block of memory.  Looking at g_strlcpy it looks like it also just works with bytes so then we could switch over using that one instead.